### PR TITLE
Core: Don't rely on java time for epoch seconds formatting

### PR DESCRIPTION
**Pull request by spinscale**
In order to be compatible with joda time, this adds an epoch seconds
formatter, that is able to parse floating point values.

However joda time discards the floating point values, but still parses
the data, where as this one is able to parse the whole value including
milliseconds.

This also allows to name the formatter 'epoch_seconds' with an s at the
end just like 'epoch_millis'.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 1]
```diff
@@ -1374,7 +1373,7 @@ public static DateFormatter forPattern(String input, Locale locale) {
             return YEAR_MONTH;
         } else if ("yearMonthDay".equals(input) || "year_month_day".equals(input)) {
             return YEAR_MONTH_DAY;
-        } else if ("epoch_second".equals(input)) {
+        } else if ("epoch_second".equals(input) || "epoch_seconds".equals(input)) {
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: If we are going to add the plural (+1), then we should add a deprecation warning for the non-plural.

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java [hunk 2]
```diff
@@ -893,8 +893,7 @@
     /*
      * Returns a formatter for parsing the seconds since the epoch
      */
-    private static final DateFormatter EPOCH_SECOND = new JavaDateFormatter("epoch_second",
-        new DateTimeFormatterBuilder().appendValue(ChronoField.INSTANT_SECONDS).toFormatter(Locale.ROOT));
+    private static final DateFormatter EPOCH_SECOND = EpochSecondsDateFormatter.INSTANCE;
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: Why do we need the constant now, we can just reference the singleton in the one place this is used?

File: server/src/main/java/org/elasticsearch/common/time/EpochSecondsDateFormatter.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+    public static DateFormatter INSTANCE = new EpochSecondsDateFormatter();
+
+    private EpochSecondsDateFormatter() {}
+
+    @Override
+    public TemporalAccessor parse(String input) {
+        try {
+            if (input.contains(".")) {
+                String[] inputs = input.split("\\.", 2);
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/time/EpochSecondsDateFormatter.java]: Can we use a precompiled regex?

File: server/src/main/java/org/elasticsearch/common/time/EpochSecondsDateFormatter.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                return Instant.ofEpochSecond(Long.valueOf(input)).atZone(ZoneOffset.UTC);
+            }
+        } catch (NumberFormatException e) {
+            throw new DateTimeParseException("invalid number", input, 0, e);
+        }
+    }
+
+    @Override
+    public DateFormatter withZone(ZoneId zoneId) {
+        return this;
```
Comment (rjernst) [inline server/src/main/java/org/elasticsearch/common/time/EpochSecondsDateFormatter.java]: This seems trappy...we should at least have a comment why this is ok, but it seems like we could drop timezones?

File: server/src/test/java/org/elasticsearch/common/time/DateFormattersTests.java [hunk 5]
```diff
@@ -56,6 +57,38 @@ public void testEpochMilliParser() {
...
+        assertThat(Instant.from(formatter.parse("1234.1")).getNano(), is(100_000_000));
+        assertThat(Instant.from(formatter.parse("1234.12")).getNano(), is(120_000_000));
+        assertThat(Instant.from(formatter.parse("1234.123")).getNano(), is(123_000_000));
+        assertThat(Instant.from(formatter.parse("1234.1234")).getNano(), is(123_400_000));
+        assertThat(Instant.from(formatter.parse("1234.12345")).getNano(), is(123_450_000));
+        assertThat(Instant.from(formatter.parse("1234.123456")).getNano(), is(123_456_000));
+        assertThat(Instant.from(formatter.parse("1234.1234567")).getNano(), is(123_456_700));
+        assertThat(Instant.from(formatter.parse("1234.12345678")).getNano(), is(123_456_780));
+        assertThat(Instant.from(formatter.parse("1234.123456789")).getNano(), is(123_456_789));
+        expectThrows(DateTimeParseException.class, () -> formatter.parse("1234.1234567890"));
```
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/common/time/DateFormattersTests.java]: We should check the exception messages with these throws

File: server/src/test/java/org/elasticsearch/common/time/DateFormattersTests.java [hunk 6]
```diff
@@ -67,6 +100,8 @@ private void assertThatSameDateTime(DateFormatter formatter, DateFormatter zoned
         String millisAsString = String.valueOf(millis);
         ZonedDateTime formatterZonedDateTime = DateFormatters.toZonedDateTime(formatter.parse(millisAsString));
         ZonedDateTime zonedFormatterZonedDateTime = DateFormatters.toZonedDateTime(zonedFormatter.parse(millisAsString));
+        logger.info(formatterZonedDateTime.toInstant());
```
Comment (rjernst) [inline server/src/test/java/org/elasticsearch/common/time/DateFormattersTests.java]: leftover from debugging?

**Review by rjernst — APPROVED** [review]
LGTM, I left a few suggestions

File: server/src/main/java/org/elasticsearch/common/time/EpochSecondsDateFormatter.java (hunk 4 above)
Comment (spinscale) [reply server/src/main/java/org/elasticsearch/common/time/EpochSecondsDateFormatter.java]: the java-time branch will create new objects with a timezone for all the date formatters (same needs to be done for the locale). I will create a separate PR for that after this one.

File: server/src/main/java/org/elasticsearch/common/time/DateFormatters.java (hunk 1 above)
Comment (spinscale) [reply server/src/main/java/org/elasticsearch/common/time/DateFormatters.java]: ok, this requires some more changes in other tests, I will add this in a follow up PR, but do not want to clutter within this PR

File: server/src/test/java/org/elasticsearch/common/time/DateFormattersTests.java (hunk 6 above)
Comment (spinscale) [reply server/src/test/java/org/elasticsearch/common/time/DateFormattersTests.java]: yes

File: server/src/test/java/org/elasticsearch/common/time/DateFormattersTests.java (hunk 5 above)
Comment (spinscale) [reply server/src/test/java/org/elasticsearch/common/time/DateFormattersTests.java]: done

**Pull request merged**